### PR TITLE
CANS-1167: performance improvements

### PR DESCRIPTION
**Pull request by LeonidMarushevskyi**
## JIRA
https://osi-cwds.atlassian.net/browse/CANS-1167

## Description

Domain and Item render only when props or state have changed. Reduce rerendering of complete Assessment component tree on each change of rating in the item.

## Tests

- [ ] I used TDD to develop the feature
- [x] I have included unit tests
- [ ] I have included new acceptance tests or modified existing ones
- [ ] I have included other tests

## Types of changes

- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (No behavioral changes)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I ran all my tests locally which were green.
- [x] My code adds no new issues to Code Climate
- [x] I ran all the linters for the project.
- [x] My code is in a stable state ready to be deployable, but not necessarily complete.
- [x] I promise on my honor as a CWDS developer to ensure I don't break the build (or bring donuts if I leave things broken), to check linters, use best practices, to leave the code in better shape than I found it, and to have a good day.

### CODE REVIEW

**Comment by LeonidMarushevskyi** [comment]
Jenkins retest this please

**Comment by wmitchell-cambria** [comment]
See my PR #666 and recently on the expand collapse for some guidance here. Instead of updating the `componentShouldUpdate` method here and using JSON stringify, we should instead make our updating functions be more careful about things like not updating references when the data doesn't change.

JSON.stringify here has its own performance issues, and it is error prone when the props cannot be stringified (if they contain circular references, for example), or when the stringified versions are not actually equivalent (function or component props).

File: app/javascript/Application/components/Assessment/Domain.js [hunk 1]
```diff
@@ -174,31 +178,33 @@ class Domain extends React.PureComponent {
...
-          />
-          {isCaregiverDomain &&
-            !this.props.disabled && (
-              <DomainCaregiverControls
-                onRemoveCaregiverDomain={this.handleRemoveCaregiverDomain}
-                onAddCaregiverDomain={this.handleAddCaregiverDomain}
-              />
-            )}
-        </ExpansionPanelDetails>
+        {isExpanded && (
```
Comment (LeonidMarushevskyi) [inline app/javascript/Application/components/Assessment/Domain.js]: Not renders domain internals if it is not expanded.

**Comment by LeonidMarushevskyi** [comment]
@wmitchell-cambria I removed everything related to componentShouldUpdate after #675. Please revisit .

**Pull request merged**